File: backend/app/schemas/strategy.py
"""Strategy Pydantic schemas for request/response validation."""

from datetime import datetime
from typing import Optional, Dict, Any
from pydantic import BaseModel, Field, ConfigDict

from app.core.strategy.types import StrategyType, StrategyStatus


# Base schema
class StrategyBase(BaseModel):
    """Base strategy schema."""
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    strategy_type: StrategyType
    parameters: Dict[str, Any] = Field(default_factory=dict)


# Schema for creating strategy
class StrategyCreate(StrategyBase):
    """Schema for creating a new strategy."""
    pass


# Schema for updating strategy
class StrategyUpdate(BaseModel):
    """Schema for updating a strategy."""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    strategy_type: Optional[StrategyType] = None
    parameters: Optional[Dict[str, Any]] = None
    status: Optional[StrategyStatus] = None


# Schema for strategy response
class StrategyResponse(StrategyBase):
    """Schema for strategy response."""
    id: int
    user_id: int
    status: StrategyStatus
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# Schema for strategy list response
class StrategyListResponse(BaseModel):
    """Schema for paginated strategy list."""
    strategies: list[StrategyResponse]
    total: int
    page: int
    page_size: int


# Schema for strategy execution request
class StrategyExecuteRequest(BaseModel):
    """Schema for executing strategy on multiple symbols."""
    symbols: list[str] = Field(..., min_length=1, description="List of stock symbols to execute strategy on")
    source: Optional[str] = Field(None, description="Source of symbols (manual, watchlist, screening)")


# Schema for signal in execution response
class SignalResponse(BaseModel):
    """Schema for a single trading signal."""
    timestamp: str
    symbol: str
    signal_type: str
    price: float
    quantity: Optional[int] = None
    reason: Optional[str] = None
    confidence: Optional[float] = None


# Schema for strategy execution response
class StrategyExecuteResponse(BaseModel):
    """Schema for strategy execution response."""
    strategy_id: int
    strategy_name: str
    symbols: list[str]
    executed_at: str
    signals: list[SignalResponse]
    total_signals: int
    total_symbols: int

File: frontend/src/types/strategy.ts
// Backend에서 사용하는 enum 값 (대문자)
export type StrategyType = 'MOMENTUM' | 'MEAN_REVERSION' | 'BREAKOUT' | 'CUSTOM'

export type StrategyStatus = 'ACTIVE' | 'INACTIVE' | 'BACKTESTING'

export interface Strategy {
  id: number
  name: string
  description?: string
  strategy_type: StrategyType  // Backend 필드명과 일치
  status: StrategyStatus
  parameters: Record<string, any>
  created_at: string
  updated_at: string
  user_id: number
  // Performance metrics (아직 구현 안됨)
  total_profit_loss?: number
  profit_loss_percent?: number
  total_trades?: number
  win_rate?: number
  sharpe_ratio?: number
  max_drawdown?: number
}

export interface CreateStrategyRequest {
  name: string
  description?: string
  strategy_type: StrategyType
  parameters: Record<string, any>
}

export interface UpdateStrategyRequest {
  name?: string
  description?: string
  strategy_type?: StrategyType
  parameters?: Record<string, any>
  status?: StrategyStatus
}

export const STRATEGY_TYPE_LABELS: Record<StrategyType, string> = {
  MOMENTUM: '모멘텀 전략',
  MEAN_REVERSION: '평균회귀 전략',
  BREAKOUT: '변동성 돌파 전략',
  CUSTOM: '커스텀 전략',
}

export const STRATEGY_STATUS_LABELS: Record<StrategyStatus, string> = {
  ACTIVE: '운영중',
  INACTIVE: '비활성',
  BACKTESTING: '백테스트중',
}

export const STRATEGY_STATUS_COLORS: Record<StrategyStatus, string> = {
  ACTIVE: 'bg-green-100 text-green-700',
  INACTIVE: 'bg-gray-100 text-gray-700',
  BACKTESTING: 'bg-blue-100 text-blue-700',
}

// API 응답 타입
export interface StrategyListResponse {
  strategies: Strategy[]
  total: number
  page: number
  page_size: number
}

export interface ExecuteStrategyResponse {
  strategy_id: number
  strategy_name: string
  symbols: string[]
  executed_at: string
  signals: Signal[]
  total_signals: number
  total_symbols: number
}

export interface Signal {
  timestamp: string
  symbol: string
  signal_type: 'BUY' | 'SELL' | 'HOLD'
  price: number
  quantity: number | null
  reason: string
  confidence: number
}
